### DISCUSSION TITLE
progressive loading for big diagrams possible ?

### DISCUSSION THREAD

**Discussion by IOR88** [xyflow/xyflow#3033] [Ideas]
Hello everyone, we have have quit big diagrams, with 100 and more elements going to few hundreds and many paths. Loading of the full diagram on bootstrap is making the app very slow, is it possible somehow to discover what is the current viewport that user is looking on and trigger a data upload handler that will load a part of the diagram ? Or maybe sb has solved similar problem with tools that react-flow has now ? What we try to achieve is to load parts of the diagram in packages, based on user activity maybe.

Best,
Igor

**Comment by namdaoduy** [comment]
It looks like you're facing a **performance problem with rendering** instead of data fetching.
It's a common problem when rendering too many DOM nodes. In this case, I suggest implementing lazy loading mechanism based on the viewport and positions of elements.

Assuming:
- You've fetched all the data for the diagrams.
- You've fetched (or calculated) positions for all nodes and edges (elements and paths).

With these pieces of information:
- Current `flowTransform`
- `width` and `height` of the current viewport
- positions of all nodes
- positions of all edges
- Rectangle area of a node
- Rectangle area of an edge

You can calculate:
- If a node is visible in the viewport
- If an edge is visible in the viewport

Then you only display these:
- Visible nodes
- Visible edges
- Edges that attach to a visible node

To improve performance, you can consider:
- Recalculate visible nodes/edges using `throttle`, since `flowTransform` updates frequently.
- Move the calculation to [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) to unblock the main thread.

---
My team solved this problem for thousands of nodes. Feel free if you need further explanation or implementation details 😃

**Reply by ajaysanthoshb** [reply]
Can you send implementation logic please , I have those above piece of informations but don't know how to relate the `flowTransform` with `width, height` of viewport.

**Reply by loggerhead** [reply]
@namdaoduy I have the exact same problem as @ajaysanthoshb. Is there any demo code or pseudocode? That would be extremely helpful. And do we still need enable `onlyRenderVisibleElements` after this kind of optimization?

**Comment by IOR88** [comment]
Thank You for support @namdaoduy we have as well one custom integration with react-flow which may causing the whole problem, we use svgjs but as I read here https://github.com/svgdotjs/svg.js/issues/740 svgjs is not so easy to integrate svgjs with virtual dom.

We use svgjs because we needed similar to drawio (https://app.diagrams.net/) feature, like very precise path creation, would You have some advice how to approach it ?

**Comment by loggerhead** [comment]
Hey guys, I have implemented the code that @namdaoduy mentioned. If anyone is still facing this problem, you can take a look at [my code](https://github.com/loggerhead/json4u/blob/main/src/lib/graph/virtual.ts) for reference. The final result after optimization looks like the GIF below. And you can try it online in [JSON For You](https://json4u.com/editor) with the [1MB-min.json](https://github.com/user-attachments/files/17265123/1MB-min.json) file.

![graph](https://github.com/user-attachments/assets/d8196ed4-d2e1-4a26-bfc4-02a89627f2fb)
